src/ch/ch_hotplug.c

```diff
@@ -0,0 +1,194 @@
+/*
+ * ch_hostdev.h: Cloud Hypervisor hotplug support
+ *
+ * Copyright Intel Corp. 2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "ch_cgroup.h"
+#include "ch_domain.h"
+#include "ch_hotplug.h"
+#include "ch_monitor.h"
+#include "ch_process.h"
+#include "domain_audit.h"
+
+#define VIR_FROM_THIS VIR_FROM_CH
+
+/**
+ * virCHDomainSelectHotplugVcpuEntities:
+ *
+ * @def: domain definition
+ * @nvcpus: target vcpu count
+ *
+ * Validate vcpu entities will be able to be enabled/disabled.
+ *
+ * Returns the true if the nvcpus can be set, false if not.
+ */
+static bool
+virCHDomainSelectHotplugVcpuEntities(virDomainDefPtr def,
+                                     unsigned int nvcpus)
+{
+    bool ret = false;
+    virDomainVcpuDefPtr vcpu;
+    unsigned int maxvcpus = virDomainDefGetVcpusMax(def);
+    unsigned int curvcpus = virDomainDefGetVcpus(def);
+    ssize_t i;
+
+    if (nvcpus > curvcpus) {
+        for (i = 0; i < maxvcpus && curvcpus < nvcpus; i++) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (vcpu->online)
+                continue;
+
+            curvcpus++;
+
+            if (curvcpus > nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    } else {
+        for (i = maxvcpus - 1; i >= 0 && curvcpus > nvcpus; i--) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (!vcpu->online)
+                continue;
+
+            curvcpus--;
+
+            if (curvcpus < nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    }
+
+    if (curvcpus != nvcpus) {
```

Comment:
I think we can move this check to the start of the function by modifying `if (nvcpus < 1 || nvcpus > maxvcpus) { error }`. The current implementation seems to be okay to hot-unplug all vcpus. I feel that we should not allow unplugging all vcpus. This check will also take care of that.

Reply:
The qemu version of this is a bit different but I did try to keep the same overall structure hence the logic around unplugging all cpus not handled here. This check's avoids adding extra inconsistency in case we had a system where `vcpu->online` is true for more vcpus than what `virDomainDefGetVcpus` gives us (the final total of cpus to be onlined at the end of the operation matches the expected value).

Maybe all we should care about is that the min and max boundaries are handled though which I'm open to so I'll update the code for just the simple check against `1` and `maxvcpus`. Cloud-Hypervisor doesn't really handle individual vcpu offlining. Letting it run and see if the vcpu structure updates after it runs corrects the previously mentioned potential inconsistencies seems like a more proactive approach anyway.

So talked myself into fixing =).